cmd/gpu_plugin/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,403 @@
+	"k8s.io/kubernetes/pkg/kubelet/apis/podresources"
+)
+
+const (
+	// ResourcePrefix is the intel resource prefix.
+	resourcePrefix = namespace + "/"
+
+	gasTSAnnotation   = "gas-ts"
+	gasCardAnnotation = "gas-container-cards"
+
+	grpcAddress    = "unix:///var/lib/kubelet/pod-resources/kubelet.sock"
+	grpcBufferSize = 4 * 1024 * 1024
+	grpcTimeout    = 5 * time.Second
+
+	retryTimeout = 1 * time.Second
+)
+
+// Errors.
+var (
+	errMultiContainer   = errors.New("multi-container allocation request")
+	errMonitoringDevice = errors.New("monitoring device is ignored")
+	errNoAnnotations    = errors.New("annotations missing from container allocate response")
+	errRetry            = errors.New("things didn't work out, but perhaps a retry will help")
+	errZeroPending      = errors.New("there are no pending pods anymore in this node")
+)
+
+type podCandidate struct {
+	pod                 *v1.Pod
+	name                string
+	allocatedContainers int
+	allocationTargetNum int
+}
+
+type resourceManager struct {
+	mutex     sync.RWMutex // for devTree updates during scan
+	devTree   dpapi.DeviceTree
+	nodeName  string
+	clientset *kubernetes.Clientset
+}
+
+func newResourceManager(enabled bool) *resourceManager {
+	if !enabled {
+		return nil
+	}
+
+	clientset := getClientset()
+
+	if clientset == nil {
+		klog.Error("no clientset available, resource manager not created")
+		return nil
+	}
+
+	rm := resourceManager{
+		nodeName:  os.Getenv("NODE_NAME"),
+		clientset: clientset,
+	}
+
+	klog.Info("GPU device plugin resource manager enabled")
+
+	return &rm
+}
+
+// reallocateWithFractionalResources runs the fractional resource logic.
+func (rm *resourceManager) reallocateWithFractionalResources(allocateResponse *pluginapi.AllocateResponse) error {
+	var err error
+	if err = reallocateInputCheck(allocateResponse); err != nil {
+		// it is better to give some gpu devices and return nil, than halt the whole cluster into unexpected admission errors
+		return nil
+	}
+
+	podCandidate, err := rm.findAllocationPodCandidate()
+	if err == errRetry {
+		klog.Warning("retrying POD resolving after sleeping")
+		time.Sleep(retryTimeout)
+		podCandidate, err = rm.findAllocationPodCandidate()
+	}
+
+	if err != nil {
+		if err != errZeroPending {
+			klog.Error("allocation candidate not found, perhaps the GPU scheduler extender is not called, err:" +
+				err.Error())
+		}
+		// it is better to give some gpu devices and return nil, than halt the whole cluster into unexpected admission errors
+		return nil
+	}
+
+	pod := podCandidate.pod
+	cards := containerCards(pod, podCandidate.allocatedContainers)
+	logPodInfo(pod, cards)
+
+	cresp := allocateResponse.ContainerResponses[0]
+	// select gpus for the allocation. the cards may differ from kubelet given deviceIDs!
+	for _, cardName := range cards {
+		rm.addDevicesAndMounts(cresp, cardName)
+	}
+	return nil
+}
+
+func reallocateInputCheck(allocateResponse *pluginapi.AllocateResponse) error {
```

Comment:
We happily ignore all errors, so no need to return them, just throw a warning to logs.

Reply:
It won't work here. The checks need to be in a separate function due to cilint related complexity requirements. Still, the reallocate needs to return nil without trying the rest of the code. Thus we need to return the error here, so that reallocate can return.

With unlimited complexity allowed for the reallocate, I could remove this function and just return from the reallocate-function, when any of the input checks fail.

Reply:
Probably this could work
```go
func isInputOk(allocateResponse *pluginapi.AllocateResponse) bool {
    if len(allocateResponse.ContainerResponses) != 1 {
        klog.Warning("multi container");
        return false;
    }
    ...
    return true;
}

func (rm *resourceManager) reallocateWithFractionalResources(allocateResponse *pluginapi.AllocateResponse) error {
}
    if !isInputOk(allocateResponse) {
        return nil;
    }
```

Reply:
sure, I can do that, if getting rid of the errors is important